python/pylibraft/pylibraft/neighbors/ivf_flat/ivf_flat.pyx

```diff
@@ -708,3 +708,141 @@ def search(SearchParams search_params,
         raise ValueError("query dtype %s not supported" % queries_dt)
 
     return (distances, neighbors)
+
+@auto_sync_handle
+def save(filename, Index index, handle=None):
+    """
+    Saves the index to file.
+
+    Saving / loading the index is experimental. The serialization format is
+    subject to change.
+
+    Parameters
+    ----------
+    filename : string
+        Name of the file.
+    index : Index
+        Trained IVF-Flat index.
+    {handle_docstring}
+
+    Examples
+    --------
+    >>> import cupy as cp
+
+    >>> from pylibraft.common import DeviceResources
+    >>> from pylibraft.neighbors import ivf_flat
+
+    >>> n_samples = 50000
+    >>> n_features = 50
+    >>> dataset = cp.random.random_sample((n_samples, n_features),
+    ...                                   dtype=cp.float32)
+
+    >>> # Build index
+    >>> handle = DeviceResources()
+    >>> index = ivf_flat.build(ivf_flat.IndexParams(), dataset, handle=handle)
+    >>> ivf_flat.save("my_index.bin", index, handle=handle)
+    """
+    if not index.trained:
+        raise ValueError("Index need to be built before saving it.")
+
+    if handle is None:
+        handle = DeviceResources()
+    cdef device_resources* handle_ = \
+        <device_resources*><size_t>handle.getHandle()
+
+    cdef string c_filename = filename.encode('utf-8')
+
+    cdef IndexFloat idx_float
+    cdef IndexInt8 idx_int8
+    cdef IndexUint8 idx_uint8
+
+    if index.active_index_type == "float32":
+        idx_float = index
+        c_ivf_flat.serialize(deref(handle_), c_filename, deref(idx_float.index))
+    elif index.active_index_type == "byte":
+        idx_int8 = index
+        c_ivf_flat.serialize(deref(handle_), c_filename, deref(idx_int8.index))
+    elif index.active_index_type == "ubyte":
+        idx_uint8 = index
+        c_ivf_flat.serialize(deref(handle_), c_filename, deref(idx_uint8.index))
+    else:
+        raise ValueError("Index dtype %s not supported" % index.active_index_type)
+
+
+@auto_sync_handle
+def load(filename, dtype, handle=None):
```

Comment:
Note that IVF-Flat index has a dtype parameter. The default `serialize` / `deserialize` functions do not save this information, therefore we need an extra user arg. (Well, actually it is saved in the numpy header of the dataset mdspan, but it would be somewhat hacky to load that information).

Alternatively, we could modify `save` / load` `to serialize dtype information as well. Do we need such convenience function?

Reply:
I think serializing this with the index makes the most sense. Then the user wouldn't have to know or even care what dtype was serialized- we can always give them back the correct index dtype.

Reply:
I have exposed a variant of the serialize function that uses a string as output. We save the dtype information + the serialized string on the Cython side.